gitlab_project_exporter/gitlab_project.py

```diff
@@ -0,0 +1,46 @@
+import logging
+from enum import IntEnum
+from typing import Never
+
+from gitlab import Gitlab
+from pydantic import BaseModel
+
+LOG = logging.getLogger(__name__)
+TIMEOUT = 10
+
+
+class MirrorStatusCode(IntEnum):
+    OK = 0
+    KO = 1
+
+
+class RemoteMirrorStatus(BaseModel):
+    mirror_id: str
+    url: str
+    status: MirrorStatusCode
+
+
+GetRemoteMirrorsStatusType = list[RemoteMirrorStatus] | list[Never]
+
+
+class GitlabProject:
+    def __init__(
+        self, project_id: str, gitlab_client: Gitlab, timeout: int = TIMEOUT
+    ) -> None:
+        self.timeout = timeout
+        self.project = gitlab_client.projects.get(project_id, timeout=self.timeout)
+
+    def get_remote_mirrors_status(self) -> GetRemoteMirrorsStatusType:
+        mirrors = self.project.remote_mirrors.list(timeout=self.timeout)
+        response = []
+        for m in mirrors:
+            status = (
+                MirrorStatusCode.KO
+                if m.last_update_started_at > m.last_successful_update_at
+                else MirrorStatusCode.OK
```

Comment:
`OK` and `KO` are confusing, and the mirror maybe in progress if it's started but not finish yet. Can we check `update_status` value?

There are some official [metrics](https://docs.gitlab.com/ee/administration/monitoring/prometheus/gitlab_metrics.html#sidekiq-metrics) for reference:

* gitlab_transaction_event_remote_mirrors_failed_total
* gitlab_transaction_event_remote_mirrors_finished_total
* gitlab_transaction_event_remote_mirrors_running_total

Reply:
I haven't found an enum for that field... Searching through the code of gitlab I find a test: 

* https://gitlab.com/gitlab-org/gitlab/-/blob/l.rosa/test-migration-env/spec/models/remote_mirror_spec.rb

with some values that may indicate error: `to_retry` or `failed`

The schema list it as a string: https://gitlab.com/gitlab-org/gitlab/-/blob/master/spec/fixtures/api/schemas/remote_mirror.json

The current logic have issues when a mirror is caught in the middle of an attempt, I do agree.

So we can look for `update_status in ['failed', ' retry] or not last_error`

`last_error` gets wiped out after mirror is in sync'ed again.

WDYT?

Reply:
we can use the same logic in model statemachine https://gitlab.com/gitlab-org/gitlab/-/blob/06cf7af7413c13c7cfd0a667e53014d0d0693280/app/models/remote_mirror.rb#L40-83

```
:none
:started
:finished
:failed
:to_retry
```

Reply:
reading the state machine code, we can just forget about last_error and concentrate in 

```
update_status in ["none", "failed", "retry"]
```

Reply:
implemented above proposal in f3d24dfb8df8ab50f7a692839b0cd4d69c89c0bc

Reply:
can we also rename `KO` to something like `NOT_OK` or `FAILED`?

Reply:
Yeah, it will look nicer. Done in adb880b588196996e526404ddbc0c0025c2dbac5